### DISCUSSION TITLE
Browser File Select vs Fetch API in File Column of Collection

### DISCUSSION THREAD

**Discussion by Mashiane** [pocketbase/pocketbase#6202] [Q&A]
Hi there

I have a File column in my collection. If I use a file input tag, which fires a file select in my UI, the file column gets updated without any issues and data is saved in my collection.

The issue is when I use a fetch to get the file from a URL and then convert it to a File Object. This is my javascript code.

`async function fetchFileAsObject(url, fileName) {
  try {
    // Fetch the file from the given URL
    const response = await fetch(url);

    // Check if the response is successful
    if (!response.ok) {
      throw new Error(`Failed to fetch the file: ${response.statusText}`);
    }

    // Read the response as a Blob
    const blob = await response.blob();

    // Create a File object from the Blob
    const file = new File([blob], fileName, { type: blob.type });

    return file;
  } catch (error) {
    console.error("Error fetching the file:", error);
    throw error;
  }
}`

const file = await fetchFileAsObject(fileUrl, fileName);

The returned file from this function just never gets saved by my app. I could be missing something.

What I want to do is that when creating user profiles on my app, they should have a default profile picture which they can later change, thus fetching an existing file and assigning it to the file / avatar column.

**Comment by Mashiane** [comment]
The fetch does return a file object

![image](https://github.com/user-attachments/assets/70e57f9e-c331-44c5-be88-75597abac6e6)

This is rather strange...

**Reply by ganigeorgiev** [reply]
Are you sure that you are submitting the `file` value correctly? Please show a code sample how you are sending the user create/update request.

**Comment by Mashiane** [comment]
Let me provide some pseudo code of what works and what does not work...

**THIS WORKS PERFECTLY**

`const fileInput = document.getElementById('fileInput');

const formData = new FormData();

// listen to file input changes and add the selected files to the form data
fileInput.addEventListener('change', function () {
    const files = fileInput.files;
    const firstFile = files[0];
    formData.append('documents', firstFile);
});

...

// upload and create new record
const createdRecord = await pb.collection('example').create(formData);`

**THIS DOES NOT WORK**

`const btnFile = document.getElementById('btnFile');

const formData = new FormData();

btnFile.addEventListener('click', function (e) {
    const firstFile = await fetchFileAsObject('http://???/oneplace.jpg', 'oneplace.jpg');
    formData.append('documents', firstFile);
});

...

// upload and create new record
const createdRecord = await pb.collection('example').create(formData);`

The difference per ser, is that in one, we use a file select (works and saves to pocketbase) and the other using fetch call (does not save). 

My apologies, I am using a javascript transpiler and thus decided to highlight the issue I am having.

**Reply by ganigeorgiev** [reply]
I'm not able to reproduce it. Here is a plain `pb_public/index.html`:

```html
<!DOCTYPE html />
<html>
<head>
    <meta charset="utf-8" />
    <title>Test</title>
</head>
<body>
    <script type="module">
        // this is using the latest JS SDK version compatible with PocketBase v0.23+
        import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@master/dist/pocketbase.es.mjs"

        const pb = new PocketBase("http://127.0.0.1:8090");

        async function fetchFileAsObject(url, fileName) {
            try {
                // Fetch the file from the given URL
                const response = await fetch(url);

                // Check if the response is successful
                if (!response.ok) {
                  throw new Error(`Failed to fetch the file: ${response.statusText}`);
                }

                // Read the response as a Blob
                const blob = await response.blob();

                // Create a File object from the Blob
                const file = new File([blob], fileName, { type: blob.type });

                return file;
            } catch (error) {
                console.error("Error fetching the file:", error);
                throw error;
            }
        }

        const formData = new FormData();

        const firstFile = await fetchFileAsObject('https://presentator.io/_app/immutable/assets/slide1.REVn5w03.png', 'oneplace.jpg');

        formData.append('documents', firstFile);

        await pb.collection("example").create(formData)
    </script>
</body>
</html>
```

You are not reporting and providing the exact code sample you are testing with and I cannot help you further. I've asked you several times in the past to try to elaborate a little more on your use case and to be more speciific but you refuse to do so and continue to treat the discussions and issues tracker as an llm chat and I'm no longer willing to engage with your posts.

As a last recommendation, make sure that:
- there are no errors with the fetch (e.g. CORS)
- you are using the correct `file` field name when submitting the file
- you are using JS SDK and PocketBase executable versions compatible with each other (PocketBase executable <= v0.22.x is compatible with JS SDK up to v0.21.5; PocketBase executable v0.23+ requires JS SDK v0.22+)
- check the logs in the _Dashboard > Logs_ and/or start the executable with `--dev` for more detailed error messages

**Comment by Mashiane** [comment]
@ganigeorgiev 

Thank you so much for your help. I guess there is something that I am missing on my side such that this file issue is not working. I will recheck also on my pockethost deployment what could be happening about the app. Above all else, my heart is full of gratitude because my app looks like this and so far it works as expected, in terms of what I conceptualized.

![image](https://github.com/user-attachments/assets/469d4a8d-d7a9-409f-8adf-3468b7793a85)

Thank you again for bearing with me. These are my current challenges.

1. I dont have any experience with linux. I am self teaching myself
2. I dont have any good experience with javascript either and are learning the basics. As mentioned earlier on, I am using a transpiler, this means for me I write almost non existent javascript code, or even css and html. With the previous code I sent (my apologies for seemingly wasting your time), I tried my best to make it readable and understandable to you because the only thing I changed was the source of the file, from file select input to fetch and the app broke.

So, I needed a way to avoid giving you this code (this is surely not what you expect) and then explaining how it worked, which will take us back and forth and waste more of your time. This has been produced by the transpiler in DEBUG mode. 

`// [1351] Sub CREATE_FORMDATA(fileField As String) As Map 
_B.create_formdata=async function(_filefield) {

var _qflds,_qexpand,_opt,_m,_firstfilename,_url;
// [1352] 	If ShowLog Then 
if (_B._showlog) {
// [1353] 		Log( {92} ) 
console.log('■[SDUIPocketBase: 1353]',"SDUIPocketBase." + _B._stablename + ".CREATE_FORMDATA");
// [1354] 	End If 
}

// [1355] 	If Schema.Size = 0 Then 
if (Object.keys(_B._schema).length==0) {
// [1356] 		BANano.Throw( {93} ) 
throw "SDUIPocketBase.CREATE_FORMDATA: " + _B._stablename + " - please execute SchemaAdd?? first to define your table schema!";
// [1357] 		Return Null 
return null;
// [1358] 	End If 
}

// [1359] 	Try 
try {
// [1360] 		Dim qflds As String = Join( {609} , flds) 
_qflds=_B.join(",",_B._flds,_B);
// [1361] 		Dim qexpand As String = Join( {610} , expand) 
_qexpand=_B.join(",",_B._expand,_B);
// [1363] 		Dim opt As Map = CreateMap() 
_opt={};
// [1365] 		opt.Put( {613} , keepalive) 
_opt["keepalive"]=_B._keepalive;
// [1366] 		If headers.Size <> 0 Then 

if (Object.keys(_B._headers).length!=0) {
// [1367] 			opt.Put( {614} , headers) 
_opt["headers"]=_B._headers;
// [1368] 		End If 
}
// [1369] 		If qflds <> {615} Then opt.Put( {616} , qflds) 
if (_qflds!="") {_opt["fields"]=_qflds;}
// [1370] 		If qexpand <> {617} Then opt.Put( {618} , qexpand) 
if (_qexpand!="") {_opt["expand"]=_qexpand;}
// [1371] 		If ShowLog Then 

if (_B._showlog) {
// [1372] 			Log( {94} ) 
console.log('■[SDUIPocketBase: 1372]',"SDUIPocketBase." + _B._stablename + ".CREATE_FORMDATA(" + _B._formdata + ")");
// [1373] 		End If 
}
// [1374] 		Dim m As Map = BANano.Await(client.RunMethod( {619} , sTableName).RunMethod( {620} , Array(formData, opt))) 
_m=await _B._client["collection"](_B._stablename)["create"](_B._formdata,_opt);
// [1375] 		Record = BANano.DeepClone(m) 
_B._record=DeepClone(_m);
// [1376] 		Dim firstFilename As String = m.Get(fileField) 
_firstfilename=_m[_filefield];
// [1377] 		Dim url As String = client.GetField( {621} ).RunMethod( {622} , Array(m, firstFilename, opt)).result 
_url=_B._client["files"]["getUrl"](_m,_firstfilename,_opt);

// [1378] 		m.Put( {623} , url) 
_m["fileurl"]=_url;
// [1379] 		Record.Put( {624} , url) 
_B._record["fileurl"]=_url;
// [1380] 		Return m 
return _m;
// [1381] 	Catch 
} catch(err) {

// [1382] 		Log(LastException) 
console.log('■[SDUIPocketBase: 1382]',err);
// [1383] 		Return Null 
return null;
// [1384] 	End Try 
}
// End Sub
};`

Again, thanks for your patience and in doing your best to explain the concepts to me as if you are explaining to a 5 year old. It helped a great deal. With your help my app works exactly as I thought and wanted it to. Nothing after all was impossible as I feared.

Thank you so much.
